File: Foundation/src/main/java/net/rickiekarp/foundation/core/types/OutputType.java
package net.rickiekarp.foundation.core.types;

/**
 * Defines the output type for a given property value.
 * @param <T> Type for the output.
 */
public interface OutputType<T> {

    OutputType<String> STRING = new OutputType<String>() {
        public String convert(String value) {
            return value;
        }

        public String toString() {
            return "OutputType.STRING";
        }
    };

    OutputType<Integer> INT = new OutputType<Integer>() {
        public Integer convert(String value) {
            return Integer.parseInt(value);
        }

        public String toString() {
            return "OutputType.INTEGER";
        }
    };

    OutputType<Boolean> BOOLEAN = new OutputType<Boolean>() {
        public Boolean convert(String value) {
            return Boolean.parseBoolean(value);
        }

        public String toString() {
            return "OutputType.STRING";
        }
    };

    /**
     * Convert the given value to a requested format.
     *
     * @param value property value
     * @return property value into requested format.
     */
    T convert(String value);
}

File: Foundation/src/main/kotlin/net/rickiekarp/foundation/core/types/OutputType.kt
package net.rickiekarp.foundation.core.types

/**
 * Defines the output type for a given property value.
 * @param <T> Type for the output.
</T> */
interface OutputType<T> {

    /**
     * Convert the given value to a requested format.
     *
     * @param value property value
     * @return property value into requested format.
     */
    fun convert(value: String): T

    companion object {

        val STRING: OutputType<String> = object : OutputType<String> {
            override fun convert(value: String): String {
                return value
            }

            override fun toString(): String {
                return "OutputType.STRING"
            }
        }

        val INT: OutputType<Int> = object : OutputType<Int> {
            override fun convert(value: String): Int {
                return Integer.parseInt(value)
            }

            override fun toString(): String {
                return "OutputType.INTEGER"
            }
        }

        val BOOLEAN: OutputType<Boolean> = object : OutputType<Boolean> {
            override fun convert(value: String): Boolean {
                return java.lang.Boolean.parseBoolean(value)
            }

            override fun toString(): String {
                return "OutputType.STRING"
            }
        }
    }
}